### DISCUSSION TITLE
Help understanding cache

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#12008] [Archive]
I'm currently evaluating Directus and in reading the configuration I see that it has caching or the option to use Redis, neither of which we have used before. Our current LAMP stack sites simply query directly against the database without any caching, so clearly we don't need extreme levels of performance, however we also do not use an API server currently so everything goes directly to the database. I suppose MySQL has a cache, but anyway, you get the picture. 

If we introduce Directus that would mean that we would move all of our SQL (business logic) into custom endpoints in Directus. This  means we are intentionally creating overhead, so I can see caching requests and results being very important, but I have questions about how the cache knows that that the underlying data has changed, especially in relational queries. 

Let's say I have an order and order lines setup and we do an API query to grab all order lines for an order along with some order information. We do this by order number to get all order lines. Next let's say we do an API PUT to update one of the order lines. How does the cache know that order line has changed and now knows to return the updated line, when we are still using the same order number? Does the cache track changes to the underlying data of anything in its cache? 

Next, let's say I want to update another order line directly in SQL in Directus because its part of another larger business logic endpoint. Do we have to let the cache know somehow that we have updated that order line or do we need to implement this as a PUT on itself so that the cache knows? OR perhaps there are cache functions in which we tell it to invalidate the cache for a particular query or set of parameters?

I see there are config parameters to invalidate the cache whenever a PUT, etc are performed. I assume this doesn't invalidate the entire cache, but instead only when the PUT directly relates to that cache entry? 

Please help me to understand how this works. This has bearing on how we would migrate to Directus and get the best performance when moving our business logic into an intentionally higher overhead architecture. 

All the best.

**Comment by rijkvanzanten** [comment]
> How does the cache know that order line has changed and now knows to return the updated line, when we are still using the same order number?

It doesn't! The cache is a little more barebones, in that it will match the http request that was made, and return early if a cached version exist. The cache auto-purge environment variable will instruct the API to clear all content caches on CrUD operations, to make sure you keep the data real time. If that realtime-ness is of less importance, you could disable that and use a lower TTL value for the cache. At the end of the day, reading data from the database directly will still be _fast_. The content cache is mostly meant for very high-traffic operations, where you do _a lot_ of read operations for the same data in rapid succession (like on a live website with a lot of public data).

> Does the cache track changes to the underlying data of anything in its cache?

No, it's not that smart. (yet!)

> Do we have to let the cache know somehow that we have updated that order line or do we need to implement this as a PUT on itself so that the cache knows?

You would either clear the cache keys from Redis, or simply wait until the TTL has passed. This totally depends on your use case though, and how accurate you need the data to be (for some projects, it's fine to have outdated information float around for a couple minutes where high-performance is more important than realtime data accuracy)

**Comment by Thermal-Shock** [comment]
Thanks as always for the great reply man. I like the "yet!" in that statement. Are there any plans in the works to have deeper control over what gets cleared when a CrUD occurs on a specific collection or item? 

Seems like it would be possible to have a really nice cache section in the app dashboard and be able to configure the cache purge on a per collection basis versus having a single global that clears for all caches for all collections.

It would also be nice to to configure filters on specific items, e.g. when a CrUD happens on /item/collection/1 we want to clear the cache for it but on item/collection/2 we don't need to. Not sure when this would be that selective but Directus shouldn't care about the reason why of course. 

To complete the picture with this in place, it could also say that if you CrUD on /item/collection/:id that another /item/collection[/:id] should also be cleared and thus would allow child updates to clear the parent cache for specific items. 

I think in general just exposing more configuration of what parts of the cache can be cleared automatically could be very powerful. 

In the meantime, I could see us writing a custom endpoint to update multiple tables in a logical transaction. How do I clear the cache myself as part of that code for specific URLs? How does Directus do it when it receives a CrUD and can I simply call that abstraction from my custom endpoint? 

All the best.

**Reply by rijkvanzanten** [reply]
> Are there any plans in the works to have deeper control over what gets cleared when a CrUD occurs on a specific collection or item?

There's no current in-works sprints to improve on this, though i'm definitely in favor of making the caching smarter and more configurable. 

> Seems like it would be possible to have a really nice cache section in the app dashboard and be able to configure the cache purge on a per collection basis versus having a single global that clears for all caches for all collections.

I agree! That being said, I do think we should aim for a setup that just works without the need for manual configuration. As soon as you start to configure some collections different from others, it might get confusing when data is cached vs when it isn't. 

> It would also be nice to to configure filters on specific items, e.g. when a CrUD happens on /item/collection/1 we want to clear the cache for it but on item/collection/2 we don't need to.

That's where it gets really complicated really fast. /item/collection/1 might have a relationships to collection `another`, which in turn means that a CrUD to /item/collection/1 should also purge the cache for `another`. Once that's the case, you also need to know whichever other endpoint might have access that particular item in `another` to ensure the cache is successfully purged. That's also the reason why the current implementation is this "brute" in saying "Just purge _everything_ to be safe". It's definitely less optimized, but reading from the database fresh that first time is often fast enough to not really be impacted by the difference. 

> I think in general just exposing more configuration of what parts of the cache can be cleared automatically could be very powerful.

I'm wondering if it should be deeply configurable, or generally just be smarter about it. If you have to configure all these rules manually, it feels like it would be very easy to mistakenly configure it "halfway", which might cause way bigger issues 🤔 

> How do I clear the cache myself as part of that code for specific URLs? How does Directus do it when it receives a CrUD and can I simply call that abstraction from my custom endpoint?

I'm not sure if the cache object is currently passed into the context of a registerEndpoint function (I don't think it is). You might be able to import it through `import { getCache } from 'directus/cache'`, but your luck may vary (haven't tried this myself)

**Reply by ithan** [reply]
Hello @rijkvanzanten! Sorry for reviving this after so long, but our team is working on a project with Directus and we're encountering issues where we cannot use Directus caching easily, since we are writing frequently to some tables that trigger invalidation for everything, making caching impossible for us. (The tables that we are writing to also have heavy reads, so we need to cache them as well, and we can't just add them to the ignore list.)

Do you have any plans to improve the granularity of the current cache?

For our current use case, defining from the data model which collections should be invalidated on write/update requests would be enough, but potentially allowing us to write custom filters for each collection to define what should be invalidated could have some benefits as well. But this is just for our use case, of course!

**Reply by rijkvanzanten** [reply]
Hey @ithan! I do have some hopes and dreams, but no concrete plans. Given that any read can contain nested relational items, it's quite tricky to know what data points exist within a given cache key. We're currently caching on the directus-query level before the database queries are executed. The best case scenario here is to revise the caching engine to cache the actual database queries themselves, and keep track of what collections are included in each database query. That way the cache can more performantly be shared between users that make identical queries [^1] and it allows us to invalidate the cache keys where certain collections are included through either FROM or JOINs. 

Setting up a custom filter per collection is a clever workaround, but not something I'm comfortable in shipping for everybody, as it requires quite a lot of insight into the exact queries you're making which would make it pretty easy to mess up for regular users.

[^1]: Permissions are included within the query itself, so as long as we include the full SQL query in the hash of the cache key different users with the same permission rules can still safely share a cache
